addons/misra.py

```diff
@@ -256,14 +256,105 @@ def getForLoopExpressions(forToken):
+            ret.append(cond.astOperand1)
+        elif cond.astOperand2.isName and cond.astOperand1.isNumber:
+            ret.append(cond.astOperand2)
+        elif cond.astOperand2.isOp and cond.astOperand1.isOp:
+            ret.extend(findCounterTokens(cond.astOperand1))
+            ret.extend(findCounterTokens(cond.astOperand2))
+        elif cond.astOperand1.isOp and cond.astOperand2.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand1))
+        elif cond.astOperand2.isOp and cond.astOperand1.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand2))
+
+    return ret
+
+
+def getFloatComparision(floatCompOperator):
+    while not floatCompOperator.isComparisonOp:
+        floatCompOperator = floatCompOperator.astParent
+    expr = hasFloatComparison(floatCompOperator)
+    if expr:
+        return True
+    return False
+
+
+def getWhileLoopExpressions(whileToken):
+    if not simpleMatch(whileToken, 'while ('):
+        return None
+    lpar = whileToken.next
+    if not lpar.astOperand2.astOperand1:
+        return None
+    rpar = lpar.link
+    counter = findCounterTokens(lpar.astOperand2)
+    if len(counter) == 0:
+        return None
+    if simpleMatch(rpar, ') {'):
+        token = rpar.next
+        while (token != rpar.next.link):
+            token = token.next
+            for counter_str in counter:
+                if token.isAssignmentOp and token.astOperand1.str == counter_str.str:
+                    return getFloatComparision(counter_str.astParent)
+                elif token.str == counter_str.str and token.astParent and token.astParent.str in {'++', '--'}:
+                    return getFloatComparision(counter_str.astParent)
+    elif simpleMatch(whileToken.previous, '} while') and simpleMatch(whileToken.previous.link.previous, 'do {'):
+        token = whileToken.previous
+        while (token != whileToken.previous.link):
+            token = token.previous
+            for counter_str in counter:
+                if token.isAssignmentOp and token.astOperand1.str == counter_str.str:
+                    return getFloatComparision(counter_str.astParent)
+                elif token.str == counter_str.str and token.astParent and token.astParent.str in {'++', '--'}:
+                    return getFloatComparision(counter_str.astParent)
+    return None
+
+
 def hasFloatComparison(expr):
     if not expr:
         return False
     if expr.isLogicalOp:
         return hasFloatComparison(expr.astOperand1) or hasFloatComparison(expr.astOperand2)
+    if expr.isArithmeticalOp:
+        if expr.astOperand1.isFloat or expr.astOperand2.isFloat:
+            return expr.astOperand1.isFloat or expr.astOperand2.isFloat
+        if expr.astOperand1.isArithmeticalOp or expr.astOperand2.isArithmeticalOp:
+            return hasFloatComparison(expr.astOperand1) or hasFloatComparison(expr.astOperand2)
     if expr.isComparisonOp:
-        # TODO: Use ValueType
```

Comment:
To me it seems like a good idea to use `ValueType`.

Reply:
I just want to know that why is it better to use valueType.type == "float" or valueType.type == "double" instead of .isFloat

Reply:
As far as I know, Token.isFloat is only true for floating point numbers. So if you have a expression that have float type I believe that Token.isFloat is False for that.

    float f = 10.0f;   // <- isFloat is True for 10.0f
    x = f;  // <- isFloat is False for f